working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+those)* is desugared to a class declaration without a primary
+constructor. An enum declaration with a primary constructor is desugared
+using the same steps. This determines the static analysis and dynamic
+semantics of a primary constructor.
+
+*In other words, there is no other semantics than the desugaring. When
+desugared, the resulting class declaration is treated exactly the same as
+it would have been if the developer had written the result of the
+desugaring step in the first place.*
+
+A compile-time error occurs if a formal parameter that contains a
+term of the form `this.v`, or `super.v` where `v` is an identifier has the
+modifier `covariant`. *For example, `required covariant int this.v` is an
+error.*
+
+Conversely, it is not an error for the modifier `covariant` to occur on
+other formal parameters of a primary constructor. *This extends the
+existing whitelist of places where `covariant` can occur.*
+
+The desugaring consists of the following steps, where _D_ is the class or
+enum declaration in the program that includes a primary constructor, and
+_D2_ is the result of desugaring. The desugaring step will delete elements
+that amount to the primary constructor, and it will add a new constructor
+_k_.
+
+Where no processing is mentioned below, _D2_ is identical to _D_. Changes
+occur as follows:
+
+The current scope of the formal parameter list of the primary constructor
+in _D_ is the current scope of the class/enum declaration *(in other words,
+the default values cannot see declarations in the class body)*.  Every
+default value in the primary constructor of _D_ is replaced by a fresh
+private name `_n`, and a constant variable named `_n` is added to the
+top-level of the current library. *(This means that we can move the
+parameter declarations including the default value without changing its
+meaning.)*
+
+Next, _k_ is a constant constructor iff the keyword `const` occurs just
+before the class name in the header of _D_, or _D_ is an `enum`
+declaration.
+
+If the class name `C` in _D_ is followed by `.id` where `id` is an
+identifier then _k_ has the name `C.id`. If it is followed by `.new` then
+_k_ has the name `C`. If the class name is not followed by `.` then _k_ has
+the name `C`.
+
+If it exists, _D2_ omits the part derived from `'.' <identifierOrNew>` that
+follows the class name in _D_.
+
+_D2_ omits the formal parameter list _L_ that follows the class name and
+possibly `.id` in _D_.
+
+The formal parameter list _L2_ of _k_ is identical to _L_, except that each
+formal parameter is processed as follows.
+
+In particular, the formal parameters in _L_ and _L2_ occur in the same
+order, and mandatory (respectively optional) positional parameters remain
+mandatory (respectively optional, with the same default value, if any), and
+named parameters preserve the name and the property of being `required`.
+
+- An initializing formal parameter *(e.g., `this.x`)* is copied from _L_ to
+  _L2_ without changes.
```

Comment:
Initializing formal parameters can have default values. If the rewrite of those into constant variables above was considered done on *D*, then it doesn't need mentioning again. If the rewrite is intended to be atomic, it needs to be mentioned here (and probably *done* here).

Reply:
Indeed, improved now.